synapse/handlers/sync.py

```diff
@@ -1631,9 +1657,17 @@ async def _get_rooms_changed(
             should_invite = non_joins[-1].membership == Membership.INVITE
             if should_invite:
                 if event.sender not in ignored_users:
-                    room_sync = InvitedSyncResult(room_id, invite=non_joins[-1])
-                    if room_sync:
-                        invited.append(room_sync)
+                    invite_room_sync = InvitedSyncResult(room_id, invite=non_joins[-1])
+                    if invite_room_sync:
+                        invited.append(invite_room_sync)
+
+            # Only bother if our latest membership in the room is knock (and we haven't
+            # been accepted/rejected in the meantime).
+            should_knock = non_joins[-1].membership == Membership.KNOCK
```

Comment:
Do we care about the room version here?

Reply:
The latest membership in the room shouldn't be knock if the event auth rules disallow knocks from entering the room state.

Reply:
Makes sense!